File: src/main/java/com/unihogsoft/scalag/vulkan/utility/VulkanObject.java
package com.unihogsoft.scalag.vulkan.utility;

/**
 * @author MarconZet
 * Created 13.04.2020
 */
public abstract class VulkanObject {
    private boolean alive;

    public VulkanObject() {
        alive = false;
    }

    public void destroy() {
        if(!alive)
            throw new IllegalStateException();
        close();
        alive = false;
    }

    protected void create() {
        if(alive)
            throw new IllegalStateException();
        init();
        alive = true;
    }

    protected abstract void init();

    protected abstract void close();

    public boolean isAlive() {
        return alive;
    }
}

File: src/main/scala/com/scalag/vulkan/utility/VulkanObject.scala
package com.scalag.vulkan.utility;

/** @author
  *   MarconZet Created 13.04.2020
  */
abstract class VulkanObject {
  protected var alive: Boolean = false

  def destroy(): Unit = {
    if (!alive)
      throw new IllegalStateException();
    close();
    alive = false;
  }

  protected def create(): Unit = {
    if (alive)
      throw new IllegalStateException();
    init();
    alive = true;
  }

  protected def init(): Unit

  protected def close(): Unit

}
